File: src/WithStrategyPattern/PaymentSystem.java
package WithStrategyPattern;

public class PaymentSystem {
    private PaymentStrategy strategy;

    // This allows the strategy to be swapped at runtime
    public void setPaymentStrategy(PaymentStrategy strategy) {
        this.strategy = strategy;
    }

    public void executePayment(int amount) {
        strategy.pay(amount);
    }
}

File: design-patterns/python/Strategy-DesignPattern/WithStrategyPattern/payment_system.py
"""Context that uses a payment strategy."""
from typing import Optional
from .payment_strategy import PaymentStrategy


class PaymentSystem:
    def __init__(self) -> None:
        self._strategy: Optional[PaymentStrategy] = None

    def set_payment_strategy(self, strategy: PaymentStrategy) -> None:
        self._strategy = strategy

    def execute_payment(self, amount: int) -> None:
        if self._strategy is None:
            raise RuntimeError("Payment strategy not set")
        self._strategy.pay(amount)
